### PR TITLE
[Merged by Bors] - First prototype for new `JsString` using UTF-16 

### PR DESCRIPTION
**Pull request by jedel1043**
I think it's time to address the elephant in the room.

This Pull Request will (hopefully!) solve part of #736.

This is a complete rewrite of `JsString`, but instead of storing `u8` bytes it stores `u16` words. The `encode!` macro (renamed to `utf16!` for simplicity) from the `const-utf16` crate allows us to create UTF-16 encoded arrays at compilation time. `JsString` implements `Deref<Target=[u16]>` to unlock the slice methods and possibly make some manipulations easier. However, we would need to create our own library of utilities for `JsString`.

### CODE REVIEW

**Comment by jedel1043** [comment]
In addition, the [utf16_literal](https://docs.rs/utf16_literal) crate should be useful to convert runtime strings from `String` to `JsString`

File: boa/src/jsstring.rs
Comment (jasonwilliams) [inline boa/src/jsstring.rs:207]: @jedel1043 what's PhantomData here and why is it used? First time I've come across it

Comment (jedel1043) [reply boa/src/jsstring.rs:207]: It's a hint to tell the compiler that dropping a `JsString` may drop `Inner`. However, I was just copying the implementation of `Rc` from the std, and after some research I realized it may not be necessary in this case, it's just needed when you have self-referential structures or lifetimes that aren't used inside the struct definition. (Reference: https://stackoverflow.com/a/58034644)
I'll still try to test if removing the `PhantomData` works without problems.

**Comment by Razican** [comment]
Hi @jedel1043! Do you think you would be able to get back to this? This would be very nice to have :)

**Comment by jedel1043** [comment]
I'm still working on it locally :)
I'll push incremental commits until I integrate the new type into all the builtins

**Comment by jedel1043** [comment]
Time to see how many tests break... 😅

**Comment by jedel1043** [comment]
### Test262 conformance changes
  
  #### VM implementation
  | Test result | main count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 87,912 | 87,912 | 0 |
  | Passed | 40,956 | 40,936 | **-20** |
  | Ignored | 21,153 | 21,153 | 0 |
  | Failed | 25,803 | 25,823 | **+20** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 46.59% | 46.56% | **-0.02%** |
  
  <details><summary><b>Fixed tests (16):</b></summary>
  
  ```
  test/built-ins/String/prototype/replaceAll/searchValue-flags-null-undefined-throws.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-null-undefined-throws.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-get-flags-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-get-flags-abrupt.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-toString-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-toString-abrupt.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-isRegExp-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-isRegExp-abrupt.js (previously Failed)
  test/built-ins/String/prototype/startsWith/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/startsWith/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/String/prototype/endsWith/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/endsWith/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/String/prototype/includes/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/includes/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-ascii.js [strict mode] (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-ascii.js (previously Failed)
  ```
  </details>
  
  <details><summary><b>Broken tests (36):</b></summary>
  
  ```
  test/language/literals/regexp/u-null-character-escape.js [strict mode] (previously Passed)
  test/language/literals/regexp/u-null-character-escape.js (previously Passed)
  test/built-ins/Array/prototype/concat/Array.prototype.concat_spreadable-string-wrapper.js [strict mode] (previously Passed)
  test/built-ins/Array/prototype/concat/Array.prototype.concat_spreadable-string-wrapper.js (previously Passed)
  test/built-ins/RegExp/S15.10.2.10_A3.1_T1.js [strict mode] (previously Passed)
  test/built-ins/RegExp/S15.10.2.10_A3.1_T1.js (previously Passed)
  test/built-ins/RegExp/character-class-escape-non-whitespace-u180e.js [strict mode] (previously Passed)
  test/built-ins/RegExp/character-class-escape-non-whitespace-u180e.js (previously Passed)
  test/built-ins/RegExp/S15.10.2.10_A4.1_T1.js [strict mode] (previously Passed)
  test/built-ins/RegExp/S15.10.2.10_A4.1_T1.js (previously Passed)
  test/built-ins/RegExp/S15.10.2.10_A4.1_T3.js [strict mode] (previously Passed)
  test/built-ins/RegExp/S15.10.2.10_A4.1_T3.js (previously Passed)
  test/built-ins/RegExp/u180e.js [strict mode] (previously Passed)
  test/built-ins/RegExp/u180e.js (previously Passed)
  test/built-ins/RegExp/lookBehind/start-of-line.js [strict mode] (previously Passed)
  test/built-ins/RegExp/lookBehind/start-of-line.js (previously Passed)
  test/built-ins/RegExp/prototype/exec/u-captured-value.js [strict mode] (previously Passed)
  test/built-ins/RegExp/prototype/exec/u-captured-value.js (previously Passed)
  test/built-ins/RegExp/prototype/Symbol.match/builtin-success-u-return-val-groups.js [strict mode] (previously Passed)
  test/built-ins/RegExp/prototype/Symbol.match/builtin-success-u-return-val-groups.js (previously Passed)
  test/built-ins/String/fromCharCode/S9.7_A3.2_T1.js [strict mode] (previously Passed)
  test/built-ins/String/fromCharCode/S9.7_A3.2_T1.js (previously Passed)
  test/built-ins/String/fromCharCode/S9.7_A3.1_T3.js [strict mode] (previously Passed)
  test/built-ins/String/fromCharCode/S9.7_A3.1_T3.js (previously Passed)
  test/built-ins/String/fromCharCode/S9.7_A3.1_T2.js [strict mode] (previously Passed)
  test/built-ins/String/fromCharCode/S9.7_A3.1_T2.js (previously Passed)
  test/built-ins/String/prototype/toLowerCase/special_casing_conditional.js [strict mode] (previously Passed)
  test/built-ins/String/prototype/toLowerCase/special_casing_conditional.js (previously Passed)
  test/built-ins/String/prototype/toLowerCase/Final_Sigma_U180E.js [strict mode] (previously Passed)
  test/built-ins/String/prototype/toLowerCase/Final_Sigma_U180E.js (previously Passed)
  test/built-ins/String/prototype/padEnd/normal-operation.js [strict mode] (previously Passed)
  test/built-ins/String/prototype/padEnd/normal-operation.js (previously Passed)
  test/built-ins/String/prototype/padStart/normal-operation.js [strict mode] (previously Passed)
  test/built-ins/String/prototype/padStart/normal-operation.js (previously Passed)
  test/annexB/built-ins/String/prototype/substr/surrogate-pairs.js [strict mode] (previously Passed)
  test/annexB/built-ins/String/prototype/substr/surrogate-pairs.js (previously Passed)
  ```
  </details>

**Comment by jedel1043** [comment]
Well, I expected to see more failed tests tbh. It could also be that we cannot parse string literals with UTF-16 codepoints; the interner automatically converts unpaired surrogates to `\uFFFD`. I'll probably fix those today or tomorrow.

On another note, the interner doesn't support UTF-16 strings, so we will need to contribute to `string-interner` the missing features.

**Comment by jedel1043** [comment]
### Test262 conformance changes
  
  #### VM implementation
  | Test result | main count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 87,912 | 87,912 | 0 |
  | Passed | 40,956 | 40,984 | **+28** |
  | Ignored | 21,153 | 21,153 | 0 |
  | Failed | 25,803 | 25,775 | **-28** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 46.59% | 46.62% | **+0.03%** |
  
  <details><summary><b>Fixed tests (38):</b></summary>
  
  ```
  test/built-ins/DataView/prototype/setUint16/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setUint16/set-values-return-undefined.js (previously Failed)
  test/built-ins/DataView/prototype/setUint8/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setUint8/set-values-return-undefined.js (previously Failed)
  test/built-ins/DataView/prototype/setInt8/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setInt8/set-values-return-undefined.js (previously Failed)
  test/built-ins/DataView/prototype/setInt16/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setInt16/set-values-return-undefined.js (previously Failed)
  test/built-ins/TypedArray/prototype/set/typedarray-arg-set-values-diff-buffer-other-type-conversions.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/set/typedarray-arg-set-values-diff-buffer-other-type-conversions.js (previously Failed)
  test/built-ins/TypedArray/prototype/set/array-arg-src-tonumber-value-conversions.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/set/array-arg-src-tonumber-value-conversions.js (previously Failed)
  test/built-ins/TypedArray/prototype/map/return-new-typedarray-conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/map/return-new-typedarray-conversion-operation.js (previously Failed)
  test/built-ins/TypedArray/prototype/fill/fill-values-conversion-operations.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/fill/fill-values-conversion-operations.js (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/DefineOwnProperty/conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/DefineOwnProperty/conversion-operation.js (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/Set/conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/Set/conversion-operation.js (previously Failed)
  test/built-ins/TypedArrayConstructors/ctors/object-arg/conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArrayConstructors/ctors/object-arg/conversion-operation.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-null-undefined-throws.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-null-undefined-throws.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-get-flags-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-get-flags-abrupt.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-toString-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-toString-abrupt.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-isRegExp-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-isRegExp-abrupt.js (previously Failed)
  test/built-ins/String/prototype/startsWith/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/startsWith/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/String/prototype/endsWith/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/endsWith/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/String/prototype/includes/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/includes/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-ascii.js [strict mode] (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-ascii.js (previously Failed)
  ```
  </details>
  
  <details><summary><b>Broken tests (10):</b></summary>
  
  ```
  test/built-ins/Array/prototype/concat/Array.prototype.concat_spreadable-string-wrapper.js [strict mode] (previously Passed)
  test/built-ins/Array/prototype/concat/Array.prototype.concat_spreadable-string-wrapper.js (previously Passed)
  test/built-ins/RegExp/prototype/exec/u-captured-value.js [strict mode] (previously Passed)
  test/built-ins/RegExp/prototype/exec/u-captured-value.js (previously Passed)
  test/built-ins/String/prototype/padEnd/normal-operation.js [strict mode] (previously Passed)
  test/built-ins/String/prototype/padEnd/normal-operation.js (previously Passed)
  test/built-ins/String/prototype/padStart/normal-operation.js [strict mode] (previously Passed)
  test/built-ins/String/prototype/padStart/normal-operation.js (previously Passed)
  test/annexB/built-ins/String/prototype/substr/surrogate-pairs.js [strict mode] (previously Passed)
  test/annexB/built-ins/String/prototype/substr/surrogate-pairs.js (previously Passed)
  ```
  </details>

**Comment by jedel1043** [comment]
### Test262 conformance changes
  
  #### VM implementation
  | Test result | main count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 88,[342](https://github.com/boa-dev/boa/runs/5352896310?check_suite_focus=true#step:10:342) | 88,342 | 0 |
  | Passed | 42,933 | 42,975 | **+42** |
  | Ignored | 21,413 | 21,413 | 0 |
  | Failed | 23,996 | 23,954 | **-42** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 48.60% | 48.65% | **+0.05%** |
  
  <details><summary><b>Fixed tests (46):</b></summary>
  
  ```
  test/built-ins/StringIteratorPrototype/next/next-iteration-surrogate-pairs.js [strict mode] (previously Failed)
  test/built-ins/StringIteratorPrototype/next/next-iteration-surrogate-pairs.js (previously Failed)
  test/built-ins/DataView/prototype/setUint16/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setUint16/set-values-return-undefined.js (previously Failed)
  test/built-ins/DataView/prototype/setUint8/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setUint8/set-values-return-undefined.js (previously Failed)
  test/built-ins/DataView/prototype/setInt8/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setInt8/set-values-return-undefined.js (previously Failed)
  test/built-ins/DataView/prototype/setInt16/set-values-return-undefined.js [strict mode] (previously Failed)
  test/built-ins/DataView/prototype/setInt16/set-values-return-undefined.js (previously Failed)
  test/built-ins/TypedArray/prototype/set/typedarray-arg-set-values-diff-buffer-other-type-conversions.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/set/typedarray-arg-set-values-diff-buffer-other-type-conversions.js (previously Failed)
  test/built-ins/TypedArray/prototype/set/array-arg-src-tonumber-value-conversions.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/set/array-arg-src-tonumber-value-conversions.js (previously Failed)
  test/built-ins/TypedArray/prototype/map/return-new-typedarray-conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/map/return-new-typedarray-conversion-operation.js (previously Failed)
  test/built-ins/TypedArray/prototype/fill/fill-values-conversion-operations.js [strict mode] (previously Failed)
  test/built-ins/TypedArray/prototype/fill/fill-values-conversion-operations.js (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/DefineOwnProperty/conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/DefineOwnProperty/conversion-operation.js (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/Set/conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArrayConstructors/internals/Set/conversion-operation.js (previously Failed)
  test/built-ins/TypedArrayConstructors/ctors/object-arg/conversion-operation.js [strict mode] (previously Failed)
  test/built-ins/TypedArrayConstructors/ctors/object-arg/conversion-operation.js (previously Failed)
  test/built-ins/String/prototype/codePointAt/return-single-code-unit.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/codePointAt/return-single-code-unit.js (previously Failed)
  test/built-ins/String/prototype/codePointAt/return-first-code-unit.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/codePointAt/return-first-code-unit.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-null-undefined-throws.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-null-undefined-throws.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-get-flags-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-get-flags-abrupt.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-toString-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-flags-toString-abrupt.js (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-isRegExp-abrupt.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/replaceAll/searchValue-isRegExp-abrupt.js (previously Failed)
  test/built-ins/String/prototype/startsWith/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/startsWith/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/String/prototype/endsWith/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/endsWith/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/String/prototype/includes/return-abrupt-from-searchstring-regexp-test.js [strict mode] (previously Failed)
  test/built-ins/String/prototype/includes/return-abrupt-from-searchstring-regexp-test.js (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-unicode.js [strict mode] (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-unicode.js (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-ascii.js [strict mode] (previously Failed)
  test/built-ins/JSON/stringify/value-string-escape-ascii.js (previously Failed)
  ```
  </details>
  
  <details><summary><b>Broken tests (4):</b></summary>
  
  ```
  test/built-ins/RegExp/dotall/with-dotall-unicode.js [strict mode] (previously Passed)
  test/built-ins/RegExp/dotall/with-dotall-unicode.js (previously Passed)
  test/built-ins/RegExp/dotall/without-dotall-unicode.js [strict mode] (previously Passed)
  test/built-ins/RegExp/dotall/without-dotall-unicode.js (previously Passed)
  ```
  </details>

**Comment by jedel1043** [comment]
I'm gonna mark this as blocked on https://github.com/Robbepop/string-interner/pull/42. In the best case scenario, it'll get merged soon.
Also, the two failing tests have to do with our (lack of) support for UTF-16 regex, so I'm gonna work on https://github.com/ridiculousfish/regress/issues/43 in the meantime.

**Comment by jedel1043** [comment]
OK, this is officially unblocked by #2147. I'll work on rebasing this today :)

**Comment by jedel1043** [comment]
Marking as ready for review since this is pretty much feature complete. It's still missing some docs adjustments but it would be good to have some reviews in the meantime.

File: boa_engine/src/builtins/function/mod.rs
Comment (jedel1043) [reply boa_engine/src/builtins/function/mod.rs:479]: No, because the parser wouldn't be capable of parsing `UTF-8` without converting it to `UTF-16` first.
Also, we would have to convert from `UTF-16` to Unicode codepoints each time the lexer has to take the next `char`, so it's easier to ask for an `u32` iterator in the first place.

**Comment by raskad** [comment]
Amost done with my review. We should probably do a benchmark on this right @jedel1043?

**Comment by jedel1043** [comment]
Yep, let me push to activate the bot

File: boa_interner/src/lib.rs
Comment (jedel1043) [inline boa_interner/src/lib.rs:67]: I'm reevaluating whether this design is worth, because maintaining alignment consumes more memory at the long run, whereas having two independent interners for each encoding consumes the same amount of memory for small and large amounts of strings.
After benchmarks I'll experiment with another design to see if the performance improves.

**Comment by jedel1043** [comment]
Hm, a segmentation fault that is only detected by benchmarks. That's really weird...
I'll redesign the interner to see if that's the cause

**Comment by Razican** [comment]
Hmmm benchmarks show a clear regression :(

**Comment by jedel1043** [comment]
> Hmmm benchmarks show a clear regression :(

Understandable, since we now have to use double the memory to store strings. That's not cache friendly at all

**Comment by raskad** [comment]
> > I was thinking about some possible optimizations, and I think we could optimize `JsString` by only storing `UTF-16` if the string has characters outside of the ASCII range, and storing only ASCII otherwise, but that would require a bit more design effort.
> 
> AFAIK [V8](https://chromium.googlesource.com/v8/v8.git/+/refs/heads/main/src/objects/string.h#105) and spidermoney does this optimization, I think this could be done after we merge this first, since this PR is already huge

@jedel1043, @HalidOdat, @Razican Since we are at the start of 0.17 now, do we want to rebase, review and merge this as is and set a goal to improve the performance until the 0.17 release?

**Comment by jedel1043** [comment]
+1 for @raskad 's suggestion, but I'm pretty much biased haha

**Comment by Razican** [comment]
I think we can merge it as-is, yes :) then improve performance before 0.17

**Comment by jedel1043** [comment]
Rebased!

File: boa_interner/src/lib.rs
Comment (RageKnify) [inline boa_interner/src/lib.rs:353]: Maybe make this a debug assert?

**Review by raskad — COMMENTED** [review]
Still have to look at the interner changes. There are some clippy doc warnigs that I have not marked additionally.

**Review by RageKnify — APPROVED** [review]
LGTM
Congrats, this change is really cool

File: boa_interner/src/raw.rs
Comment (RageKnify) [inline boa_interner/src/raw.rs:40]: Need to update the comments to the correct struct name, here and a few more comments
```suggestion
    /// Creates a new [`RawInterner`] with the specified capacity.
```

File: boa_engine/src/builtins/function/mod.rs
Comment (Razican) [inline boa_engine/src/builtins/function/mod.rs:507]: Shouldn't it be generic to `u16` iterators?

File: boa_engine/Cargo.toml
Comment (Razican) [inline boa_engine/Cargo.toml:56]: Are these 3 new dependencies really needed? Adding more dependencies increases compile times, and we might not need all of the ones we are using.

File: boa_interner/src/lib.rs
Comment (Razican) [inline boa_interner/src/lib.rs:71]: This seems to be fixed already, right? At least in the clippy repo.

File: boa_interner/src/lib.rs
Comment (Razican) [inline boa_interner/src/lib.rs:192]: Each line should be up to 100 characters long, but as long as possible. This seems a bit chaotic. Seems to happen in all documentation comments in this file.

File: boa_interner/src/lib.rs
Comment (Razican) [inline boa_interner/src/lib.rs:72]: We might want to add examples in the documentation, right? Or maybe create an issue to track this.

File: boa_interner/Cargo.toml
Comment (Razican) [inline boa_interner/Cargo.toml:20]: Is this different to our `utf16!()` macro?

File: boa_engine/src/syntax/ast/node/identifier/mod.rs
Comment (Razican) [inline boa_engine/src/syntax/ast/node/identifier/mod.rs:4]: I would combine both imports.

File: boa_engine/src/symbol.rs
Comment (Razican) [inline boa_engine/src/symbol.rs:308]: We can probably use `map_or_else()` here for better readability.

File: boa_engine/src/builtins/function/mod.rs
Comment (jedel1043) [reply boa_engine/src/builtins/function/mod.rs:507]: Not really, since we would need to convert any source code to UTF-16 before passing it to the parser. On the other hand, we can convert both UTF-16 and UTF-8 strings to u32 codepoint iterators without allocating additional space.

File: boa_engine/Cargo.toml
Comment (jedel1043) [reply boa_engine/Cargo.toml:56]: I'll remove `itertools`, because it introduces a lot of functions we don't really use and it increases compilation times considerably.

`sptr` and `static_assertions` are essential to ensure we're not introducing UB on the `string` module. `sptr` takes care of our pointer provenance, which ensures ptr to int casts won't cause UB in the future. `static_assertions` ensures `COMMON_STRINGS` is not empty, which would cause UB since we're assuming the first string on that array is our default value.

File: boa_interner/src/lib.rs
Comment (jedel1043) [reply boa_interner/src/lib.rs:71]: Not really. If I remove that, and run `cargo clippy`, the output is:

```
error: these where clauses contain repeated elements
   --> boa_interner/src/lib.rs:194:12
    |
194 |         C: From<&'a str> + From<&'b [u16]>,
```

Comment (jedel1043) [reply boa_interner/src/lib.rs:71]: Opened https://github.com/rust-lang/rust-clippy/issues/9626, will change the link to this.

File: boa_interner/Cargo.toml
Comment (jedel1043) [reply boa_interner/Cargo.toml:20]: Oops, forgot to replace this 😅

File: boa_interner/src/lib.rs
Comment (jedel1043) [reply boa_interner/src/lib.rs:192]: Yeah, let me readjust the lengths

File: boa_interner/src/lib.rs
Comment (jedel1043) [reply boa_interner/src/lib.rs:72]: I'll create an issue, since this crate didn't have examples before this PR.

File: boa_engine/Cargo.toml
Comment (jedel1043) [reply boa_engine/Cargo.toml:56]: `static_assertions` is also a dependency we already use in `boa_interner`.

Comment (Razican) [reply boa_engine/Cargo.toml:56]: Sounds good :)

File: boa_interner/src/lib.rs
Comment (Razican) [reply boa_interner/src/lib.rs:67]: Performance of this new implementation is hitting us significantly, everything that can be done in the future to improve this will be great :)

File: boa_engine/src/string.rs
Comment (Razican) [inline boa_engine/src/string.rs:103]: I would personally move this slice to a submodule, and add a line or two of documentation. It occupies a lot of space at the top of this file.

File: boa_engine/src/string.rs
Comment (Razican) [inline boa_engine/src/string.rs:552]: I'm guessing this is the "maximum length of a string in the "common" list, right? Maybe this should be explicit in a documentation line.

**Review by Razican — APPROVED** [review]
Looks very good :) It's a pitty the performance decrease, but we can merge this now and then spend some time enhancing the performance.

Check my comments and see if they make sense, before merging

File: boa_engine/src/string.rs
Comment (jedel1043) [reply boa_engine/src/string.rs]: Done!

**Pull request closed without merging**
